src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java

```diff
@@ -0,0 +1,440 @@
+/*
+ * Copyright (c) 2024 Snowflake Computing Inc. All rights reserved.
+ */
+
+package net.snowflake.ingest.streaming.internal;
+
+import static net.snowflake.ingest.streaming.internal.DataValidationUtil.checkFixedLengthByteArray;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.math.RoundingMode;
+import java.nio.charset.StandardCharsets;
+import java.time.ZoneId;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+import net.snowflake.ingest.utils.ErrorCode;
+import net.snowflake.ingest.utils.SFException;
+import net.snowflake.ingest.utils.Utils;
+import org.apache.parquet.schema.GroupType;
+import org.apache.parquet.schema.LogicalTypeAnnotation;
+import org.apache.parquet.schema.LogicalTypeAnnotation.DecimalLogicalTypeAnnotation;
+import org.apache.parquet.schema.LogicalTypeAnnotation.ListLogicalTypeAnnotation;
+import org.apache.parquet.schema.LogicalTypeAnnotation.MapLogicalTypeAnnotation;
+import org.apache.parquet.schema.LogicalTypeAnnotation.StringLogicalTypeAnnotation;
+import org.apache.parquet.schema.LogicalTypeAnnotation.TimeLogicalTypeAnnotation;
+import org.apache.parquet.schema.LogicalTypeAnnotation.TimestampLogicalTypeAnnotation;
+import org.apache.parquet.schema.PrimitiveType;
+import org.apache.parquet.schema.Type;
+import org.apache.parquet.schema.Type.Repetition;
+
+/** Parses a user Iceberg column value into Parquet internal representation for buffering. */
+class IcebergParquetValueParser {
+
+  /**
+   * Parses a user column value into Parquet internal representation for buffering.
+   *
+   * @param value column value provided by user in a row
+   * @param type Parquet column type
+   * @param stats column stats to update
+   * @param insertRowsCurrIndex Row index corresponding the row to parse (w.r.t input rows in
+   *     insertRows API, and not buffered row)
+   * @return parsed value and byte size of Parquet internal representation
+   */
+  static ParquetBufferValue parseColumnValueToParquet(
+      Object value,
+      Type type,
+      RowBufferStats stats,
+      ZoneId defaultTimezone,
+      long insertRowsCurrIndex) {
+    Utils.assertNotNull("Parquet column stats", stats);
+    float estimatedParquetSize = 0F;
+    estimatedParquetSize += ParquetBufferValue.DEFINITION_LEVEL_ENCODING_BYTE_LEN;
```

Comment:
1. checkout the comment on this constant; we are now going to have repetition level so does that change the size estimation logic here?
2. Note that FDN tables are ALSO adding support for structured data types; arguably we'll have to add support in this SDK for them too. Our changes need to treat "iceberg data types" and "structured data type support" as separate but related concepts and not tie them up too tightly. That is, keep the door open for FDN data types + structured data type nesting.

Reply:
Added repetition level encoding size estimation.